content/guides/typescript-for-solid.mdx

```diff
@@ -0,0 +1,586 @@
+const [count, setCount] = createSignal(0);
+const [name, setName] = createSignal("");
+```
+
+In this example, TypeScript understands the types as `number` and `string`.
+This means that `count` and `name` directly receive the types `Accessor<number>` and `Accessor<string>`, respectively, without the `| undefined` tag.
+
+### Context
+
+Just as signals utilize `createSignal<T>`, context uses `createContext<T>`, which is parameterized by the type `T` of the context's value:
+
+```typescript
+type Data = { count: number; name: string };
+const dataContext = createContext<Data>();
+```
+
+`dataContext` is understood as `Context<Data | undefined>`. 
+Invoking `useContext(dataContext)` mirrors this type, returning `Data | undefined`.
+The `| undefined` arises when context is not defined in the component's ancestors, which leads `useContext` to yield `undefined`.
+
+Much like what was seen the [default values](#default-values) in signals, you can avoid the `| undefined` in the type through explicity type definitions to avoid redudancy:
+
+```typescript
+const dataContext = createContext({ count: 0, name: "" });
+```
+
+Here, TypeScript determines that `dataContext` is `Context<count: number, name: string>` -- equivalent to `Context<Data>` without the `| undefined`. 
+
+A common approach is forming a [factory function](https://www.typescriptlang.org/docs/handbook/decorators.html#decorator-factories) to generate a context's value. By using TypeScripts [`ReturnType`](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#return-type-annotations), the return type of this function can be used to type this context:
+
+```typescript
+export const makeCountNameContext = (initialCount = 0, initialName = "") => {
+  const [count, setCount] = createSignal(initialCount);
+  const [name, setName] = createSignal(initialName);
+  return [
+    { count, name },
+    { setCount, setName },
+  ] as const;
+};
+
+type CountNameContextType = ReturnType<typeof makeCountNameContext>;
+export const CountNameContext = createContext<CountNameContextType>();
+```
+
+`CountNameContextType` corresponds to the result of `makeCountNameContext`:
+
+```typescript
+[
+  {readonly count: Accessor<number>, readonly name: Accessor<string>},
+  {readonly setCount: Setter<number>, readonly setName: Setter<string>}
+]
+```
+
+To retrieve the context, using `useCountNameContext` which has aa type signature of `() => CountNameContextType | undefined`.
+
+In scenarios where you may wish to avoid `undefined` as a possible type, you can asser that the context will always be present:
+
+```typescript
+export const useCountNameContext = () => useContext(CountNameContext)!;
+```
+
+**Note:** A more secure route is to supply a default value to `createContext` to make sure that context is perpetually defined.
+
+### Components
+
+#### The basics
+
+By default, components in Solid use the generic `Component<P>` type, where `P` represents the prop's type that is typically an object.
```

Comment:
```suggestion
By default, components in Solid use the generic `Component<P>` type, where `P` represents the props' type that is an object.
```

I don't think we should imply that `props` could be anything other than a `Record<string, unknown>`.